packages/adapters/storage/src/postgres.rs

```diff
@@ -1252,37 +1217,62 @@ impl Postgres {
     }
 }
 
-async fn take_blocks_until_limit(
-    mut stream: BoxStream<'_, std::result::Result<tables::DBCompressedFuelBlock, sqlx::Error>>,
-    max_cumulative_bytes: u32,
-) -> Result<Vec<CompressedFuelBlock>> {
-    let mut blocks = vec![];
+// Helper function to count additional blocks until the buildup threshold is reached or the stream is exhausted.
+async fn count_remaining_blocks(
+    stream: &mut BoxStream<'_, std::result::Result<tables::DBCompressedFuelBlock, sqlx::Error>>,
+    buildup_detection_threshold: u32,
+) -> Result<u32> {
+    let mut count = 0u32;
+    while count < buildup_detection_threshold {
+        if stream.try_next().await?.is_some() {
+            count += 1;
+        } else {
+            break;
+        }
+    }
+    Ok(count)
+}
 
-    let mut total_bytes = 0;
+async fn take_consecutive_blocks(
+    mut stream: BoxStream<'_, std::result::Result<tables::DBCompressedFuelBlock, sqlx::Error>>,
+    target_cumulative_bytes: u32,
+    block_buildup_threshold: u32,
+) -> Result<(Vec<CompressedFuelBlock>, Option<bool>)> {
+    let mut consecutive_blocks = Vec::new();
+    let mut cumulative_bytes = 0;
+    let mut total_blocks_count = 0;
     let mut last_height: Option<u32> = None;
-    while let Some(val) = stream.try_next().await? {
-        let data_len = val.data.len();
-        if total_bytes + data_len > max_cumulative_bytes as usize {
+
+    while cumulative_bytes < target_cumulative_bytes {
+        let Some(db_block) = stream.try_next().await? else {
             break;
-        }
+        };
 
-        let block = CompressedFuelBlock::try_from(val)?;
+        total_blocks_count += 1;
+        let data_len = db_block.data.len() as u32;
+        let block = CompressedFuelBlock::try_from(db_block)?;
         let height = block.height;
-        total_bytes += data_len;
-
-        blocks.push(block);
 
-        match &mut last_height {
-            Some(last_height) if height != last_height.saturating_add(1) => {
+        // Check if the block is consecutive
+        if let Some(prev_height) = last_height {
+            if height != prev_height.saturating_add(1) {
+                // A gap is detected. Break out without adding this block.
                 break;
             }
-            _ => {
-                last_height = Some(height);
-            }
         }
+        last_height = Some(height);
+        consecutive_blocks.push(block);
+        cumulative_bytes += data_len;
     }
 
-    Ok(blocks)
+    if cumulative_bytes < target_cumulative_bytes {
+        total_blocks_count += count_remaining_blocks(&mut stream, block_buildup_threshold).await?;
+        let buildup_detected = total_blocks_count >= block_buildup_threshold;
+        Ok((consecutive_blocks, Some(buildup_detected)))
+    } else {
+        // If target bytes are reached without encountering a gap, no buildup detection is needed.
+        Ok((consecutive_blocks, None))
```

Comment:
Curious, why is it not needed? It is not clear from the function name and return types=D